src/clr-angular/utils/drag-and-drop/providers/draggable-state-registrar.ts

```diff
@@ -0,0 +1,44 @@
+/*
+ * Copyright (c) 2016-2018 VMware, Inc. All Rights Reserved.
+ * This software is released under MIT license.
+ * The full license information can be found in LICENSE in the root directory of this project.
+ */
+import {Injectable} from "@angular/core";
+import {DomAdapter} from "../../dom-adapter/dom-adapter";
+import {ClrDragEvent} from "../interfaces/drag-event";
+
+
+@Injectable()
+export class ClrDraggableStateRegistrar<T> {
```

Comment:
I'm still trying to understand what this does, but I can tell you for sure the name `ClrDraggableStateRegistrar` doesn't help at all. I have no idea what the "state" of a draggable is, or why you'd have to register it.

Reply:
I am not fond of this name too, but this is what I have come up with tentatively. The goal of this service is to capture the state that the `draggable` element was in so that we could apply it to the `draggable-ghost` element. The captured state means draggable's BoundingClientRect, ComputeStyles, and the event that draggable element was captured in.

Maybe, I should also add in a comment explaining what this service is about too.

Reply:
I've used `ClrDraggableSnapshot`.  The `register()` method has become `capture()`. And the `unregister()` has become `discard()`.